### PR TITLE
gitter: Maintain a list of threads and create only one thread per room

### PR DESCRIPTION
**Pull request by meetmangukiya**
Closes https://github.com/errbotio/err-backend-gitter/issues/19

### CODE REVIEW

File: gitter.py [hunk 1]
```diff
@@ -370,8 +375,7 @@ def query_room(self, room):
     def send_message(self, mess):
         super().send_message(mess)
         log.debug("bf body = %s" % mess.body)
-        body = self.md.convert(mess.body)  # strips the unsupported stuff.
-        log.debug("af body = %s" % body)
```
Comment (meetmangukiya) [inline gitter.py]: unrelated change

File: gitter.py [hunk 2]
```diff
@@ -214,7 +214,7 @@ class GitterBackend(ErrBot):
     This is the Gitter backend for errbot.
     """
 
-    def __init__(self, config):
+    def __init__(self, config, joined_rooms=[]):
```
Comment (zoni) [inline gitter.py]: Mutable arguments in a function definition are [a no-no in Python](http://python-guide-pt-br.readthedocs.io/en/latest/writing/gotchas/#mutable-default-arguments).

Also, since a bot starts out not joined to any room, there's no need to make this settable via the constructor. So you leave this off, and just do `self._joined_rooms = []` in the body of the function.

File: gitter.py [hunk 3]
```diff
@@ -239,6 +239,8 @@ def __init__(self, config):
                              'Accept': 'application/json'}
         self.bot_identifier = self._get_bot_identifier()
 
+        self.joined_rooms = joined_rooms
```
Comment (zoni) [inline gitter.py]: Lets make this private, since things outside the class should never mess with it (so `self._joined_rooms` instead).

File: gitter.py [hunk 4]
```diff
@@ -297,9 +299,12 @@ def background():
                 else:
                     log.debug('Received keep-alive on %s', room.name)
 
-        t = threading.Thread(target=background)
-        t.daemon = True
-        t.start()
+        if room._uri not in self.joined_rooms:
```
Comment (zoni) [inline gitter.py]: You need to make sure `self.joined_rooms` is never accessed simultaneously. For an example, see how `threading.Lock()` is used inside `errbot/backends/irc.py`.

Comment (meetmangukiya) [inline gitter.py]: done

File: gitter.py [hunk 5]
```diff
@@ -297,9 +300,14 @@ def background():
...
-        t.daemon = True
-        t.start()
+        if room._uri not in self._joined_rooms:
+            t = threading.Thread(target=background)
+            t.daemon = True
+            t.start()
+            self._joined_rooms.append(room._uri)
+        else:
+            with self._joined_rooms_lock:
+                log.info("Already joined the room.")
```
Comment (zoni) [inline gitter.py]: The thread lock (`with self._joined_rooms_lock:`) is in the wrong place here. It needs to guard access and modification of `self._joined_rooms` (which happens in `if room._uri not in self._joined_rooms` and `self._joined_rooms.append(room._uri)`.

Comment (meetmangukiya) [inline gitter.py]: done.

File: gitter.py [hunk 6]
```diff
@@ -297,9 +300,14 @@ def background():
...
 
-        t = threading.Thread(target=background)
-        t.daemon = True
-        t.start()
+        if room._uri not in self._joined_rooms:
+            t = threading.Thread(target=background)
+            t.daemon = True
+            t.start()
+            with self._joined_rooms_lock:
+                self._joined_rooms.append(room._uri)
```
Comment (zoni) [inline gitter.py]: This is still incorrect. You're still accessing `self._joined_rooms` above, without the lock held. I'll modify it to show the fix.

Comment (zoni) [inline gitter.py]: There, see https://github.com/errbotio/err-backend-gitter/pull/20/commits/881e56a5083662b39c6f53279552a8ca5cc2a171.
